### PR TITLE
DDLS-217 Add date and route to User when pre registration details are validated

### PR DESCRIPTION
**Pull request by iqpalm**
## Purpose
There are a few ways to register to the service. Each of these processes will require the user to validate their details against the Pre-registration table, but at different points compare with each other.

To help us easier identify which process a user has used to register and if they have validated their details against the pre-registration table we will add 2 fields to the user table to populate this data.

Fixes DDLS-217

## Approach
1. Added two fields with pre registration validated date and registration route to dd user table
2. Added new fields to User class on api side
3. Added new fields to User class on client side
4. Set route and pre register validated date for self registration together with updated test
5. If the route is not self registering then the route will be set depending on the logged in user's role
6. Setting the pre register validated date at correct point for the remaining routes, co-deputy, admin and org admin.
7. Setup tests to test the correct status and date are set for each route.

## Learning
_Any tips and tricks, blog posts or tools which helped you. Plus anything notable you've discovered about DigiDeps_

## Checklist
- [X] I have performed a self-review of my own code
- [ ] I have updated documentation (Confluence/ADR/tech debt doc) where relevant
- [X] I have added tests to prove my work
- [ ] The product team have approved these changes
- [ ] I have checked my work for potential security issues and refered to the [OWASP top 10](https://owasp.org/www-project-top-ten/)

### Frontend
- [ ] I have run an in-browser accessibility test (e.g. WAVE, Lighthouse)
- [ ] There are no deprecated CSS classes noted in the profiler
- [ ] Translations are used and the profiler doesn't identify any missing
- [ ] Any links or buttons added are screen reader friendly and contextually complete
- [ ] If adding GA events, I have updated or [checked](docs/runbooks/GOOGLE-ANALYTICS.md) the existing category or label values

### CODE REVIEW

File: api/app/src/Service/UserService.php [hunk 1]
```diff
@@ -39,6 +39,15 @@ public function addUser(User $loggedInUser, User $userToAdd, $data)
 
         $userToAdd->recreateRegistrationToken();
         $userToAdd->setCreatedBy($loggedInUser);
+
+        if ($loggedInUser->isLayDeputy()) {
```
Comment (Raffers) [inline api/app/src/Service/UserService.php:43]: You can clean this up a bit more using a `match` - https://www.php.net/manual/en/control-structures.match.php

File: client/app/src/Entity/User.php [hunk 2]
```diff
@@ -78,6 +78,12 @@ class User implements UserInterface, DeputyInterface, PasswordAuthenticatedUserI
     public const ACTIVATE_TOKEN_EXPIRE_HOURS = 48;
     public const PASSWORD_TOKEN_EXPIRE_HOURS = 1;
 
+    public const SELF_REGISTER = 'SELF_REGISTER';
+    public const ADMIN_INVITE = 'ADMIN_INVITE';
+    public const ORG_ADMIN_INVITE = 'ORG_ADMIN_INVITE';
+    public const CO_DEPUTY_INVITE = 'CO_DEPUTY_INVITE';
+    public const UNKNOWN_REGISTRATION_ROUTE = 'UNKNOWN';
```
Comment (Raffers) [inline client/app/src/Entity/User.php:85]: Are we going to populate all fields in the table? 
If not then why do we have an unknown value?
If we do plan to populate this field for all users, where is the migration script to update these?

File: client/app/tests/phpunit/Controller/CoDeputyControllerTest.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
+<?php
```
Comment (Raffers) [inline client/app/tests/phpunit/Controller/CoDeputyControllerTest.php:1]: Why are we testing an insertion into the DB, do we not trust the DB to do a correct insertion? 
Do we have a tests on the endpoint `/selfregister/verifycodeputy` that would sufficiently cover this.

File: client/app/tests/phpunit/Controller/UserControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
+<?php
```
Comment (Raffers) [inline client/app/tests/phpunit/Controller/UserControllerTest.php:1]: The same comments have been made on `CoDeputyControllerTest`

Comment (iqpalm) [reply client/app/tests/phpunit/Controller/UserControllerTest.php]: There were no tests in place on the client side, so I am creating a couple of tickets to get them created/refactored to be proper unit tests.

File: client/app/tests/phpunit/Controller/CoDeputyControllerTest.php (hunk 3 above)
Comment (iqpalm) [reply client/app/tests/phpunit/Controller/CoDeputyControllerTest.php]: There were no tests in place on either the client or API side, so I am creating a couple of tickets to get them created/refactored to be proper unit tests.

File: client/app/src/Entity/User.php (hunk 2 above)
Comment (Raffers) [reply client/app/src/Entity/User.php:85]: Have just seen in the migrations file, that unknown will be the default. So this is required.

**Review by Raffers — APPROVED** [review]

**Pull request merged**